Update tutorial for command buffer re-recording

charles-lunarg opened the pull request:
Means the code now calls recordCommandBuffer every frame instead of ahead of time

Many changes to the structure of chapters 14 & 15, add chapter 16 that introduces the
concept of frames in flight.

Primary focus was shifting everything away from 'swapchain image count' to
MAX_FRAMES_IN_FLIGHT. This has caused a lot of chapter 15 to need to be rewritten.
Such as: Introducing fences alongside semaphores and not later; Waiting on a fence
at the start of the frame before introducing acquireNextImage; consolidating the
Frames In Flight concepts to apply to command buffers, fences, and semaphores at
the same time.

Chapter 14 saw command buffer allocation reduced to 1 at a time. This allows the
concept of Frames in Flight to not need introduction before having a triangle
drawing on the screen.

Chapter 16 now covers the topic of Frames in Flight in whole, pulling it out of 
Chapter 15 so it can focus on the main render loop & introducing synchronization.

charles-lunarg committed:
Update tutorial for command buffer re-recording

Means the code now calls recordCommandBuffer every frame instead of ahead of time

Many changes to the structure of chapters 14 & 15, with small changes to subsequent
chapters.

Primary focus was shifting everything away from 'swapchain image count' to
MAX_FRAMES_IN_FLIGHT. This has caused a lot of chapter 15 to need to be rewritten.
Such as: Introducing fences alongside semaphores and not later; Waiting on a fence
at the start of the frame before introducing acquireNextImage; consolidating the
Frames In Flight concepts to apply to command buffers, fences, and semaphores at
the same time.

Chapter 14 saw command buffer allocation reduced to 1 at a time. This allows the
concept of Frames in Flight to not need introduction before having a triangle
drawing on the screen.

catleeball committed:
Update link to Ray Tracing book

It looks like the repo https://github.com/RayTracing/InOneWeekend now points to https://github.com/RayTracing/raytracing.github.io . Updated the link to reflect this.

charles-lunarg committed:
Update introduction to semaphores and fences

Greatly improve the descriptions of semaphores and fences before their
introduction into the code. Provide examples with psuedo code.

By using max_frames_in_flight command buffers & semaphores, we dont
need to keep track of previously sumitted frames' fences and wait
on them "just in case". This removes a lot of the confusion I had
when I first was trying to understand the vulkan update loop.

Overv committed:
Merge pull request #253 from catleeball/patch-1

Update link to Ray Tracing book

Talon1024 committed:
Update 04_Swap_chain_recreation.md

data -> date

Overv committed:
Merge pull request #254 from Talon1024/patch-1

Update 04_Swap_chain_recreation.md

SaschaWillems commented:
All the changes and additions look and read great. Imo this adds a lot to the tutorial and fills in a lot of the missing parts.

gpx1000 commented on en/03_Drawing_a_triangle/03_Drawing/02_Rendering_and_presentation.md:
```diff
@@ -19,46 +19,162 @@ void drawFrame() {
+discuss timeline semaphores. Further mention of the term semaphore exclusively
+refers to binary semaphores.
+
+A semaphore is either unsignaled or signaled. It begins life as unsignaled. The
```
this is technically closer in definition to the idea of a mutual exclusion semaphore (mutex).  The difference is to say a semaphore has a count where the count can be either 0 or some number.  If it's 0 then it's "unsignaled" and available.  While if it's non-zero then it's signaled and if you want exclusive access to it, then you must wait for the count to go back to zero.
A common way I like to think of it is think about the use case of restricting memory.  You can have as many reads as you want, but you need to wait for all the reads to finish before a write.

charles-lunarg replied:
Hmm I see where that's coming from, and is consistent with my understanding of the general concept of semaphores in computer science (like what was taught in my OS class). I'm not sure bringing in a count would help here, since we don't have multiple waiters on a binary semaphore. 

How could this be improved? I am unsure how better to describe the binary semaphore without dropping the CS sync literature book on the reader.

gpx1000 replied:
I would suggest something along the lines of this:

A binary semaphore can exist in one of two states; signaled and unsignaled.  As the name implies, this binary state is derived in much the same way that a classical semaphore behaves.  The semaphore owns a value known as a count and will signal when the count is non zero.  An example of this behavior is to think of thread memory protection.  Multiple threads can read from memory at the same time, however, each write must be protected from reading at the same time.  This core principal holds true for Vulkan Semaphores, in our example, we choose a very simple use case of a single count which produces a mutual exclusive lock.  In our case, the count is simply 0 or 1, and thus the semaphore is either signaled or not.

charles-lunarg replied:
I still feel like this brings in a lot of unnecessary information. Describing the general CS semaphore brings in a lot of other details, like the count and memory protection from threaded writes.

gpx1000 replied:
I'm sorry am a little confused, vulkan semaphores are semaphores with a count.  Or am I completely mistaken.  Which has happened before.

charles-lunarg replied:
Vulkan Binary semaphores to me are just a single boolean. Vulkan Timeline semaphores are unlike the classical semaphore description since the uint64_t is always increasing.

So I don't think you are confused, I just see adding descriptions about vulkan binary semaphores being like classical semaphores with a count to only help those who already know about the classical computer science semaphore.

gpx1000 replied:
That makes sense to me.  Your position is also closer to what the standard advocates.  I think what you have is right as is.  I think technically, it'd be better if we went back in time and advocated that the naming be mutex (mutual exclusion semaphore) as opposed to semaphore and timeline semaphore as semaphore.  Although timeline doesn't allow decrementing the count as a semaphore does those connections should make it easier for reusing the terms without confusion; also timeline allows global access like a semaphore while a mutex is restricted access like the Vulkan semaphore...
I do also wonder if it makes sense to deprecate talking about fences and semaphores and just use timeline semaphores?
I digress, I think it's right as is.  The PR looks good to me.

charles-lunarg replied:
Once timeline semaphores can interact with the swapchain, we could start advocating for that.
I agree that the naming is confusing, hard to not want to go back in time an do a lot of things different in vulkan.

gpx1000 replied:
:) isn't that the truth!

gpx1000 reviewed (commented):
Great job on this!  Very clear explanation; code looks clean and works here.
I'm suggesting a minor tweak in the semaphore description.  Otherwise LGTM;

Commit:
Mention std::clamp in explanatory comment instead of min/max

Overv committed:
Merge pull request #256 from ralphtheninja/std-clamp

Mention std::clamp in explanatory comment instead of min/max

ZaOniRinku committed:
Fix minor typo in french translation

Overv committed:
Merge pull request #258 from ZaOniRinku/master

Fix minor typo in french translation

ZaOniRinku committed:
Add full name for vkBindXMemory functions

Overv committed:
Merge pull request #259 from ZaOniRinku/master

Add full name for vkBindXMemory functions in french translation

charles-lunarg committed:
Remove accidental code changes

Un-comment vkDeviceWaitIdle and remove resizing of imagesInFlight (which was removed)

charles-lunarg committed:
Use uint32_t instead of size_t for currentFrame

charles-lunarg committed:
Address typos and fixup changes for re-recording command buffers PR

* Add description of fences needing explicit resetting while semaphores are automatic
* Elaborate on why 2 frames in flight are chosen

charles-lunarg committed:
Fix deadlock in resizing from resetting the fence too early

Because acquiring the swapchain image index may cause drawFrame to return early,
it was possible to cause the next vkWaitForFences to deadlock. By delaying fence
resetting till after acquiring, it prevents the deadlock.

thesusian committed:
Fixed getMaxUsableSampleCount() formatting

thesusian committed:
Chnage createFrameBuffers to createFramebuffers

Chnage createFrameBuffers to createFramebuffers because it's written as "createFramebuffers" in the source code

Overv committed:
Merge pull request #262 from thesusian/patch-2

Change createFrameBuffers to createFramebuffers

Overv committed:
Merge pull request #261 from thesusian/patch-1

Fixed getMaxUsableSampleCount() formatting

SaschaWillems committed:
Updated links in the sidebar

- Removed no-longer updated quick reference
- Update link to chunked Vulkan 1.2 spec
- Added link to the Khronos Vulkan Guide

SaschaWillems committed:
Updated links to chunked Vulkan 1.2 spec

Rubo3 committed:
Add Arch Linux support

Added Arch Linux support, plus a "`make`" formatting fix.

Overv committed:
Merge pull request #264 from Rubo3/patch-1

Add Arch Linux support

SaschaWillems committed:
Changed links according to QA

Use Vulkan 1.3 spec

SaschaWillems committed:
Updating 1.2 to 1.3 spec links

Overv committed:
Merge pull request #263 from SaschaWillems/update-links

Update links
